### PR TITLE
security: validate client-supplied trace IDs

### PR DESCRIPTION
**Pull request by ankurs**
## Summary
- Add `validateTraceID` helper: max 128 chars, printable ASCII only (0x20-0x7E), strips control/unicode bytes
- Called in `SetTraceIdWithValue` (after gRPC metadata extraction) and `UpdateTraceId` (before storing)
- If all characters are invalid, falls back to generating a new UUID
- Protects structured logs, Sentry/Rollbar, and OTEL span attributes from log injection and DoS

## Test plan
- [x] `TestValidateTraceID` — table-driven: normal, truncation, stripping, edge cases
- [x] `TestSetTraceId_ValidatesMetadataTraceID` — dirty metadata sanitized
- [x] `TestSetTraceId_TruncatesLongMetadataTraceID` — 200 chars → 128
- [x] `TestUpdateTraceId_ValidatesInput` — dirty input sanitized
- [x] `TestUpdateTraceId_AllInvalidFallsBackToGenerated` — generates new UUID
- [x] `make test`, `make lint`, `make doc` all pass

## Summary by CodeRabbit

* **New Features**
  * Added ability to obtain the updated context along with the resolved trace ID.
  * Added a configurable trace-ID validator hook so callers can supply custom validation or disable it.

* **Improvements**
  * Trace IDs are automatically sanitized (printable ASCII only, truncated to 128 chars).
  * Reduced default async-notification limit from 1000 to 20 and clarified concurrency semantics.

* **Documentation**
  * Updated README links and API docs to reflect these changes.

* **Tests**
  * Added unit tests covering trace-ID sanitization and validator customization.

### CODE REVIEW

File: notifier/notifier.go [hunk 1]
```diff
@@ -604,10 +606,36 @@ func GetTraceId(ctx context.Context) string {
...
+	return b.String()
+}
+
 // UpdateTraceId force updates the traced id to provided id
 // if no trace id is found then it will create one and update the context
 // You should use the context returned by this function instead of the one passed
 func UpdateTraceId(ctx context.Context, traceID string) context.Context {
+	traceID = validateTraceID(traceID)
 	if traceID == "" {
 		return SetTraceId(ctx)
```
[the remark this answers is not in the document]
Comment (ankurs) [reply notifier/notifier.go:656]: This is intentional — when the caller's explicit ID is fully invalid (all non-printable), falling back to `SetTraceId` is correct because it follows the priority chain: existing context → gRPC metadata → OTEL span → generate UUID. The metadata re-adoption is the right behavior since the original `UpdateTraceId` call was for a *different* trace ID source (proto field), not the same metadata.

File: notifier/notifier_test.go [hunk 2]
```diff
@@ -231,3 +232,69 @@ func TestSetTraceIdWithValue_SetsOTELAttribute(t *testing.T) {
...
+	tests := []struct {
+		name  string
+		input string
+		want  string
+	}{
+		{"empty stays empty", "", ""},
+		{"normal ASCII passes through", "abc-123-def", "abc-123-def"},
+		{"UUID passes through", "550e8400-e29b-41d4-a716-446655440000", "550e8400-e29b-41d4-a716-446655440000"},
+		{"truncated at 128 chars", strings.Repeat("a", 200), strings.Repeat("a", 128)},
+		{"newlines stripped", "abc\ndef\rghi", "abcdefghi"},
```
[the remark this answers is not in the document]
Comment (ankurs) [reply notifier/notifier_test.go:246]: Fixed — added non-ASCII UTF-8 test case with accented, CJK, and emoji characters.

File: notifier/notifier.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (ankurs) [reply notifier/notifier.go:656]: Fixed — `UpdateTraceId` now uses `strings.TrimSpace(traceID) == \"\"` to match `SetTraceIdWithValue` behavior.

File: notifier/notifier.go [hunk 3]
```diff
@@ -554,6 +552,10 @@ func SetTraceIdWithValue(ctx context.Context) (context.Context, string) {
 			traceID = strings.Join(id, ",")
 		}
 	}
+	// Sanitize client-supplied trace ID to prevent log injection and DoS.
+	if traceIDValidator != nil {
+		traceID = traceIDValidator(traceID)
+	}
```
[the remark this answers is not in the document]
Comment (ankurs) [reply notifier/notifier.go:558]: Pre-existing trace IDs in context were already validated when stored via `SetTraceId`/`UpdateTraceId`. Re-validating on every read would add overhead to the hot path for no benefit.

File: notifier/notifier.go [hunk 4]
```diff
@@ -604,11 +606,52 @@ func GetTraceId(ctx context.Context) string {
...
+var traceIDValidator = validateTraceID
+
+// SetTraceIDValidator sets a custom trace ID validation function.
+// The function receives a raw trace ID and must return the sanitized version.
+// Return an empty string to trigger automatic trace ID generation.
+// Set to nil to disable validation entirely (not recommended).
+// Must be called during init — not safe for concurrent use.
+func SetTraceIDValidator(fn func(string) string) {
+	traceIDValidator = fn
+}
```
[the remark this answers is not in the document]
Comment (ankurs) [reply notifier/notifier.go:621]: Fixed — updated docstring to say "triggers the standard trace ID resolution flow" instead of "triggers automatic generation".

File: notifier/README.md [hunk 5]
```diff
@@ -192,25 +194,43 @@ func SetServerRoot(path string)
...
 SetTraceHeaderName sets the header name for trace id default is x\-trace\-id
 
+<a name="SetTraceIDValidator"></a>
+## func [SetTraceIDValidator](<https://github.com/go-coldbrew/errors/blob/main/notifier/notifier.go#L618>)
+
+```go
+func SetTraceIDValidator(fn func(string) string)
+```
+
+SetTraceIDValidator sets a custom trace ID validation function. The function receives a raw trace ID and must return the sanitized version. Return an empty string to trigger automatic trace ID generation. Set to nil to disable validation entirely \(not recommended\). Must be called during init — not safe for concurrent use.
```
[the remark this answers is not in the document]
Comment (ankurs) [reply notifier/README.md]: Fixed — README updated via `make doc` to reflect the corrected docstring.

File: notifier/notifier_test.go [hunk 6]
```diff
@@ -231,3 +232,103 @@ func TestSetTraceIdWithValue_SetsOTELAttribute(t *testing.T) {
...
+	for _, tc := range tests {
+		t.Run(tc.name, func(t *testing.T) {
+			got := validateTraceID(tc.input)
+			if got != tc.want {
+				t.Errorf("validateTraceID(%q) = %q, want %q", tc.input, got, tc.want)
+			}
+		})
+	}
+}
+
```
[the remark this answers is not in the document]
Comment (ankurs) [reply notifier/notifier_test.go:264]: The early-return path only fires when a trace ID is already in context, meaning it was already validated when stored. Low value to add this test.

File: notifier/notifier.go [hunk 7]
```diff
@@ -604,11 +606,53 @@ func GetTraceId(ctx context.Context) string {
...
 }
 
+// traceIDValidator is the active trace ID validation function.
+// Defaults to validateTraceID. Set to nil to disable validation.
+var traceIDValidator = validateTraceID
+
+// SetTraceIDValidator sets a custom trace ID validation function.
+// The function receives a raw trace ID and must return the sanitized version.
+// Returning an empty string triggers the standard trace ID resolution flow
+// (OTEL span trace ID → gRPC metadata → generate UUID), not direct generation.
```
[the remark this answers is not in the document]
Comment (ankurs) [reply notifier/notifier.go]: Fixed — corrected to \`existing ctx → gRPC metadata → OTEL span trace ID → generate UUID\`.

File: notifier/notifier.go [hunk 8]
```diff
@@ -604,11 +606,53 @@ func GetTraceId(ctx context.Context) string {
...
+	var b strings.Builder
+	b.Grow(len(id))
+	for i := 0; i < len(id); i++ {
+		c := id[i]
+		if c >= 0x20 && c <= 0x7E {
+			b.WriteByte(c)
+		}
+	}
+	return b.String()
+}
```
[the remark this answers is not in the document]
Comment (ankurs) [reply notifier/notifier.go:646]: Truncating before stripping is intentional — it bounds the work upfront to at most 128 bytes of scanning. The edge case (128+ bytes of pure control chars followed by valid ASCII) is extremely unlikely in practice and would indicate a malicious or broken client. Generating a fresh UUID in that case is the correct behavior.
